tests/integ/sagemaker/workflow/test_workflow.py

```diff
@@ -1310,3 +1320,192 @@ def test_caching_behavior(
+            f"/code/{hash_files_or_dirs([DATA_DIR + '/framework_processor_data'])}"
+        )
+
+        run_procs = []
+        for step in definition["Steps"]:
+            for input_obj in step["Arguments"]["ProcessingInputs"]:
+                if input_obj["InputName"] == "entrypoint":
+                    s3_uri = input_obj["S3Input"]["S3Uri"]
+                    run_procs.append(s3_uri)
+
+                    # verify runproc.sh prefix is different from code artifact prefix
+                    assert Path(s3_uri).parent != source_dir_tar_prefix
+
+        # verify all the run_proc.sh artifact paths are distinct
+        assert len(run_procs) == len(set(run_procs))
+
+        expected_source_dir_tar = (
+            f"{pipeline_name}"
+            f"/code/{hash_files_or_dirs([DATA_DIR + '/framework_processor_data'])}/sourcedir.tar.gz"
+        )
+        expected_query_step_artifact = (
+            f"{pipeline_name}/"
+            f"code/{hash_files_or_dirs([DATA_DIR + '/framework_processor_data/query_data.py'])}/"
+            f"query_data.py"
+        )
+
+        prepare_step_script = data_processor._generate_framework_script("preprocess.py")
+        expected_prepare_step_artifact = (
+            f"{pipeline_name}/" f"code/{hash_object(prepare_step_script)}/runproc.sh"
+        )
+
+        split_step_script = data_processor._generate_framework_script("train_test_split.py")
+        expected_split_step_artifact = (
+            f"{pipeline_name}/" f"code/{hash_object(split_step_script)}/runproc.sh"
+        )
+
+        eval_step_script = sk_processor._generate_framework_script("evaluate.py")
+        expected_eval_step_artifact = (
+            f"{pipeline_name}/" f"code/{hash_object(eval_step_script)}/runproc.sh"
+        )
+
+        expected_prefix = f"{pipeline_name}/code"
+        s3_code_objects = pipeline_session.list_s3_files(
+            bucket=default_bucket, key_prefix=expected_prefix
+        )
+
+        # verify all distinct artifacts were uploaded
+        assert expected_source_dir_tar in s3_code_objects
+        assert expected_query_step_artifact in s3_code_objects
+        assert expected_prepare_step_artifact in s3_code_objects
+        assert expected_split_step_artifact in s3_code_objects
+        assert expected_eval_step_artifact in s3_code_objects
+
+        # verify runprocs contain the correct commands
+        prepare_step_runproc = S3Downloader.read_file(
+            f"s3://{default_bucket}/{expected_prepare_step_artifact}"
+        )
+        assert "python preprocess.py" in prepare_step_runproc
+
+        split_step_runproc = S3Downloader.read_file(
+            f"s3://{default_bucket}/{expected_split_step_artifact}"
+        )
+        assert "python train_test_split.py" in split_step_runproc
+
+        eval_step_runproc = S3Downloader.read_file(
+            f"s3://{default_bucket}/{expected_eval_step_artifact}"
+        )
+        assert "python evaluate.py" in eval_step_runproc
```

Comment:
In integ test, we'd better invoke a pipeline execution and check if the execution status (e.g. completed/succeeded, failed) is as expected. Otherwise, we usually put it into the unit test.

For this test, I feel we'd be better make it a integ test and trigger the execution to make sure it works E2E.

However, in this test, we add 4 processing steps into a pipeline, so the execution can take roughly 4 *4 = 16mins, which is quite time consuming. 
Can we reduce the size of this test? In other words, are all 4 steps are distinct and representing different cases?
If we're not able to reduce the size of this test and we invoke a pipeline execution here, we'd better mark the test as a slow test

Reply:
Technically, the non FrameworkProcessor steps are covered in other tests, and we can reduce the original customer problem example to just 2 FrameworkProcessor steps instead of 3. 

In this PR, the consensus with @navaj0 was to focus on pipeline definition correctness to address the specific bug, and we can add much more detailed E2E use cases for FrameworkProcessors, SparkProcessors, etc. in future improvements. But, 2 step pipelines are simple enough and execution time is still relatively short, so I reduced the steps to 2 in both tests and added the full executions for both.